cpu/native/socket_zep/socket_zep.c

```diff
@@ -408,58 +222,351 @@ static void _send_zep_hello(socket_zep_t *dev)
+    uint8_t *out = zepdev->snd_buf;
+    uint16_t chksum = 0;
+
+    out += _zep_hdr_fill(zepdev, (void *)out, bytes);
+
+    /* make sure we are not overflowing the TX buffer */
+    if (out + bytes > zepdev->snd_buf + sizeof(zepdev->snd_buf)) {
+        return -ENOBUFS;
+    }
+
+    for (unsigned i = 0; i < n; i++) {
+        memcpy(out, iolist->iol_base, iolist->iol_len);
+        chksum = ucrc16_calc_le(iolist->iol_base, iolist->iol_len,
+                                UCRC16_CCITT_POLY_LE, chksum);
+        out += iolist->iol_len;
+        iolist = iolist->iol_next;
+    }
+    chksum = byteorder_htols(chksum).u16;
+    memcpy(out, &chksum, sizeof(chksum));
+
+    zepdev->snd_len = (out + 2) - zepdev->snd_buf;
+
+    return bytes;
+}
+
+static int _request_transmit(ieee802154_dev_t *dev)
+{
+    socket_zep_t *zepdev = dev->priv;
+
+    dev->cb(dev, IEEE802154_RADIO_INDICATION_TX_START);
+
+    int res = real_write(zepdev->sock_fd, zepdev->snd_buf, zepdev->snd_len);
+
+    dev->cb(dev, IEEE802154_RADIO_CONFIRM_TX_DONE);
+
+    if (res < 0) {
+        return res;
+    }
+
+    return 0;
+}
+
+static int _confirm_transmit(ieee802154_dev_t *dev, ieee802154_tx_info_t *info)
+{
+    (void) dev;
+
+    /* TODO: handle ACK */
+    if (info) {
+        info->status = TX_STATUS_SUCCESS;
+    }
+
+    return 0;
+}
+
+int _len(ieee802154_dev_t *dev)
+{
+    size_t size;
+    socket_zep_t *zepdev = dev->priv;
+
+    int res = real_ioctl(zepdev->sock_fd, FIONREAD, &size);
+    if (res < 0) {
+        DEBUG("socket_zep::len: error reading FIONREAD: %s", strerror(errno));
+        return res;
+    }
+
+    return size;
+}
+
+static int _read(ieee802154_dev_t *dev, void *buf, size_t max_size,
+                          ieee802154_rx_info_t *info)
+{
+    int res;
+    socket_zep_t *zepdev = dev->priv;
+
+    DEBUG("socket_zep::read: reading up to %u bytes into %p\n", max_size, buf);
+
+    res = real_read(zepdev->sock_fd, buf, max_size);
+
+    DEBUG("socket_zep::read: got %d bytes\n", res);
+
+    if (res <= 0) {
+        return res;
+    }
+
+    zep_hdr_t *tmp = (zep_hdr_t *)buf;
+
+    if ((tmp->preamble[0] != 'E') || (tmp->preamble[1] != 'X')) {
+        DEBUG("socket_zep::read: invalid ZEP header\n");
+        res = -EINVAL;
+        goto out;
+    }
+
+    if (tmp->version != 2) {
+        DEBUG("socket_zep::read: unsupported ZEP version %u\n", tmp->version);
+        res = -EINVAL;
+        goto out;
+    }
+
+    switch (((zep_v2_ack_hdr_t *)tmp)->type) {
+    case ZEP_V2_TYPE_DATA: {
+        zep_v2_data_hdr_t *zep = (zep_v2_data_hdr_t *)tmp;
+
+        if (info) {
+            info->lqi = zep->lqi_val;
+            info->rssi = -IEEE802154_RADIO_RSSI_OFFSET;
+        }
+
+        if (_dst_not_me(zepdev, zep + 1)) {
+            goto out;
+        }
+
+        size_t size = zep->length - 2;
+        memmove(buf, zep + 1, size);
+        res = size;
+        break;
+    }
+    case ZEP_V2_TYPE_ACK: {
+        break;
```

Comment:
you should pass this ACK to the upper layer if you want the SubMAC to take care of retransmissions. Alternatively, you can implement this here and use caps to indicate that the ACK's are managed, but probably the former will be easier

Reply:
or is this a ZEP internal?

Reply:
Ah so I can just pass on the L2 ACK payload and don't have to distinguish between Data frames and Ack frames in the driver at all?

Reply:
I don't think that these are supposed to be used in tandem to L2 ACKs. I believe more (since they were only introduced with V2 of ZEP) that they are a ZEP internal. So rather handle them separately, in case we need them later for their actual purpose.

Reply:
> Ah so I can just pass on the L2 ACK payload and don't have to distinguish between Data frames and ACK frames in the driver at all?

This is correct. The SubMAC will take care of it

> I don't think that these are supposed to be used in tandem to L2 ACKs. I believe more (since they were only introduced with V2 of ZEP) that they are a ZEP internal (as always: ZEP is rather sparsely documented). So rather handle them separately, in case we need them later for their actual purpose.

I don't know how ZEP works. In fact, do ZEP retransmissions exists? Can these frame get lost somehow?

Reply:
> ZEP retransmissions exists?

ZEP just wraps a IEEE 802.15.4 frame inside a UDP packet, so it should be able to do everything a real IEEE 802.15.4 network can do.

> Can these frame get lost somehow?

If we simulate a network with ZEP dispatcher they [can get lost](https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/zep_dispatch/topology.c#L213), just like in a real network. :wink:

Reply:
> I don't know how ZEP works. In fact, do ZEP retransmissions exists? Can these frame get lost somehow?

Nobody knows, I think... the only doc that can be found on this is the packet format in the Wireshark code and Wiki. Since ZEP could, in theory, be sent in normal Internet traffic, I think having retransmissions in ZEP can make sense, if you want to keep the ZEP layer somewhat transparent from 802.15.4.

> If we simulate a network with ZEP dispatcher they [can get lost](https://github.com/RIOT-OS/RIOT/blob/master/dist/tools/zep_dispatch/topology.c#L213), just like in a real network. wink

I think in that case, the IEEE 802.15.4 frame should rather be dropped, not the ZEP message. Yes this implies also the ZEP message to be dropped, so semantics, I know, but a crucial difference when it comes to the layer retransmissions are happening on ;-).